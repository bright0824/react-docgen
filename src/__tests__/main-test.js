/*
 *  Copyright (c) 2015, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */

/*global jest, describe, beforeEach, it, expect*/

jest.disableAutomock();

describe('main', () => {
  var docgen, ERROR_MISSING_DEFINITION;

  beforeEach(() => {
    docgen = require('../main');
    ({ERROR_MISSING_DEFINITION} = require('../parse'));
  });

  function test(source) {
    it('parses with default resolver/handlers', () => {
      var docs = docgen.parse(source);
      expect(docs).toEqual({
        displayName: 'ABC',
        description: 'Example component description',
        methods: [],
        props: {
          foo: {
            type: {
              name: 'bool',
            },
            defaultValue: {
              computed: false,
              value: 'true',
            },
            description: 'Example prop description',
            required: false,
          },
        },
      });
    });

    it('parses with custom handlers', () => {
      var docs = docgen.parse(source, null, [
        docgen.handlers.componentDocblockHandler,
      ]);
      expect(docs).toEqual({
        description: 'Example component description',
      });
    });
  }

  describe('React.createClass', () => {
    test(`
      var React = require("React");
      var PropTypes = React.PropTypes;

      var defaultProps = {
        foo: true,
      };
      var propTypes =  {
        /**
         * Example prop description
         */
        foo: PropTypes.bool
      };

      /**
       * Example component description
       */
      var Component = React.createClass({
        displayName: 'ABC',
        propTypes,
        getDefaultProps: function() {
          return defaultProps;
        }
      });
      module.exports = Component
    `);
  });

  describe('Class definition', () => {
    test(`
      const React = require("React");
      const PropTypes = React.PropTypes;

      const defaultProps = {
        foo: true,
      };
      const propTypes =  {
        /**
         * Example prop description
         */
        foo: PropTypes.bool
      };

      /**
       * Example component description
       */
      export default class Component extends React.Component {
        static propTypes = propTypes;
        // ...
      }
      Component.defaultProps = defaultProps;
      Component.displayName = 'ABC';
    `);
  });

  describe('Stateless Component definition: ArrowFunctionExpression', () => {
    test(`
      import React, {PropTypes} from "React";

      const defaultProps = {
        foo: true,
      };
      const propTypes =  {
        /**
         * Example prop description
         */
        foo: PropTypes.bool
      };

      /**
        * Example component description
        */
      let Component = props => <div />;
      Component.displayName = 'ABC';
      Component.defaultProps = defaultProps;
      Component.propTypes = propTypes;

      export default Component;
    `);
  });

  describe('Stateless Component definition: FunctionDeclaration', () => {
    test(`
      import React, {PropTypes} from "React";

      const defaultProps = {
        foo: true,
      };
      const propTypes =  {
        /**
         * Example prop description
         */
        foo: PropTypes.bool
      };

      /**
      * Example component description
      */
      function Component (props) {
        return <div />;
      }

      Component.displayName = 'ABC';
      Component.defaultProps = defaultProps;
      Component.propTypes = propTypes;

      export default Component;
    `);
  });

  describe('Stateless Component definition: FunctionExpression', () => {
    test(`
      import React, {PropTypes} from "React";

      const defaultProps = {
        foo: true,
      };
      const propTypes =  {
        /**
         * Example prop description
         */
        foo: PropTypes.bool
      };

      /**
      * Example component description
      */
      let Component = function(props) {
        return React.createElement('div', null);
      }

      Component.displayName = 'ABC';
      Component.defaultProps = defaultProps;
      Component.propTypes = propTypes;

      export default Component;
    `);
  });

  describe('Stateless Component definition', () => {
    it('is not so greedy', () => {
      const source = `
        import React, {PropTypes} from "React";

        /**
        * Example component description
        */
        let NotAComponent = function(props) {
          let HiddenComponent = () => React.createElement('div', null);

          return 7;
        }

        NotAComponent.displayName = 'ABC';
        NotAComponent.defaultProps = {
            foo: true
        };

        NotAComponent.propTypes = {
          /**
          * Example prop description
          */
          foo: PropTypes.bool
        };

        export default NotAComponent;
      `;

      expect(() => docgen.parse(source)).toThrowError(ERROR_MISSING_DEFINITION);
    });
  });
});
