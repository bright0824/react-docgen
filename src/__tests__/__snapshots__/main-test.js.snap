exports[`main fixtures processes component "component_1.js" without errors 1`] = `
Object {
  "description": "The is a component to test the document generation",
  "displayName": "Component",
  "methods": Array [],
}
`;

exports[`main fixtures processes component "component_2.js" without errors 1`] = `
Object {
  "description": "",
  "displayName": "button",
  "methods": Array [
    Object {
      "docblock": null,
      "modifiers": Array [
        "static",
      ],
      "name": "displayName",
      "params": Array [],
      "returns": null,
    },
    Object {
      "docblock": null,
      "modifiers": Array [
        "static",
      ],
      "name": "defaultProps",
      "params": Array [],
      "returns": null,
    },
  ],
  "props": Object {
    "type": Object {
      "defaultValue": Object {
        "computed": false,
        "value": "\"primary\"",
      },
    },
  },
}
`;

exports[`main fixtures processes component "component_3.js" without errors 1`] = `
Object {
  "description": "",
  "methods": Array [],
  "props": Object {
    "style": Object {
      "description": "",
      "required": false,
      "type": Object {
        "name": "object",
      },
    },
  },
}
`;

exports[`main fixtures processes component "component_4.js" without errors 1`] = `
Object {
  "description": "",
  "methods": Array [],
  "props": Object {
    "child": Object {
      "description": "",
      "required": true,
      "type": Object {
        "computed": true,
        "name": "shape",
        "value": "Child.propTypes",
      },
    },
    "extendedChild": Object {
      "description": "",
      "required": true,
      "type": Object {
        "name": "shape",
        "value": Object {
          "adopted": Object {
            "name": "bool",
            "required": true,
          },
        },
      },
    },
    "something": Object {
      "description": "",
      "required": true,
      "type": Object {
        "name": "string",
      },
    },
  },
}
`;

exports[`main fixtures processes component "component_5.js" without errors 1`] = `
Object {
  "description": "",
  "methods": Array [],
  "props": Object {
    "children": Object {
      "description": "",
      "required": true,
      "type": Object {
        "name": "string",
      },
    },
    "onClick": Object {
      "description": "",
      "required": false,
      "type": Object {
        "name": "func",
      },
    },
    "style": Object {
      "defaultValue": Object {
        "computed": false,
        "value": "{}",
      },
      "description": "",
      "required": false,
      "type": Object {
        "name": "object",
      },
    },
  },
}
`;

exports[`main fixtures processes component "component_6.js" without errors 1`] = `
Object {
  "description": "",
  "methods": Array [],
  "props": Object {
    "children": Object {
      "description": "",
      "required": true,
      "type": Object {
        "name": "string",
      },
    },
    "onClick": Object {
      "description": "",
      "required": false,
      "type": Object {
        "name": "func",
      },
    },
    "style": Object {
      "defaultValue": Object {
        "computed": false,
        "value": "{}",
      },
      "description": "",
      "required": false,
      "type": Object {
        "name": "object",
      },
    },
  },
}
`;

exports[`main fixtures processes component "component_7.js" without errors 1`] = `
Object {
  "description": "The is a component to test the document generation",
  "displayName": "Component",
  "methods": Array [],
}
`;

exports[`main fixtures processes component "component_8.js" without errors 1`] = `
Object {
  "description": "",
  "methods": Array [
    Object {
      "docblock": null,
      "modifiers": Array [],
      "name": "onChangeSlider",
      "params": Array [
        Object {
          "name": "[min, max]",
          "optional": undefined,
          "type": null,
        },
      ],
      "returns": null,
    },
  ],
}
`;
